lib/stream.js

```diff
@@ -117,26 +121,20 @@ const applyEncoding = (contents, encoding) => encoding === 'buffer' ? new Uint8A
 
 // Transforms replace `childProcess.std*`, which means they are not exposed to users.
 // However, we still want to wait for their completion.
-const waitForOriginalStreams = (originalStreams, spawned, waitForStream) => originalStreams
-	.filter((stream, index) => stream !== spawned.stdio[index])
-	.map(stream => waitForStream(stream));
+const waitForOriginalStreams = (originalStreams, spawned, streamInfo) =>
+	originalStreams.map((stream, index) => stream === spawned.stdio[index]
+		? undefined
+		: waitForStream(stream, index, streamInfo));
 
 // Some `stdin`/`stdout`/`stderr` options create a stream, e.g. when passing a file path.
 // The `.pipe()` method automatically ends that stream when `childProcess` ends.
 // This makes sure we wait for the completion of those streams, in order to catch any error.
-const waitForCustomStreamsEnd = (stdioStreamsGroups, waitForStream) => stdioStreamsGroups.flat()
-	.filter(({type, value}) => type !== 'native' && !isStandardStream(value))
-	.map(({value}) => waitForStream(value));
-
-// Wraps `finished(stream)` to handle the following case:
-//  - When the child process exits, Node.js automatically calls `childProcess.stdin.destroy()`, which we need to ignore.
-//  - However, we still need to throw if `childProcess.stdin.destroy()` is called before child process exit.
-const onFinishedStream = async ([originalStdin], exitPromise, stream) => {
-	await Promise.race([
-		...(stream === originalStdin ? [exitPromise] : []),
-		finished(stream, {cleanup: true}),
-	]);
-};
+const waitForCustomStreamsEnd = (stdioStreamsGroups, streamInfo) => stdioStreamsGroups.flatMap((stdioStreams, index) => stdioStreams
+	.filter(({value}) => isStream(value) && !isStandardStream(value))
+	.map(({type, value}) => waitForStream(value, index, streamInfo, {
+		isSameDirection: type === 'generator',
+		stopOnExit: type === 'native',
```

Comment:
That fixes https://github.com/sindresorhus/execa/issues/825.

`type: 'native'` related to the Node.js streams described in #825. Before, they were not passed to `waitForStream()`, which handles any errors from streams. Now, it is passed, but stops listening for errors once the process exits. That's because Node.js does not destroy streams passed as options to `child_process.spawn()` once the process has ended.